README
== YubiKey Key Storage Module (YK-KSM)

The YubiKey Key Storage Module (YK-KSM) provides a AES key storage
facility for use with a YubiKey validation server.

The YK-KSM is intended to be run on a locked-down server.

This separation allows third parties to keep tight control of the AES
keys for their YubiKeys, but at the same time allow external
validation servers (e.g., Yubico's) to validate OTPs from these
YubiKeys.

The YK-KSM was designed to work with the YubiKey validation server:

https://developers.yubico.com/yubikey-val/

Documentation is in doc/.

=== Development

To create a tarball you must have a recent versions of "help2man",
"git2cl" and the perl module DBI.


NEWS
* Version 1.16 (unreleased)

  * Add support for proper HTTP status codes

  * Add support for JSON response

* Version 1.15 (released 2014-09-30)

  * Fix boolean SQL portability issue for Oracle.

  * Add self-testing through Travis.

  * Update some URLs.

* Version 1.14 (released 2013-09-18)

  * Add Oracle support via OCI.

  * The README is included in the tarball now.

* Version 1.13 (released 2013-04-17)

  * Updated release procedure, project moved from Google Code to GitHub.

* Version 1.12 (released 2013-02-05)

  * Added COPYING file.

* Version 1.11 (released 2013-01-31)

  * Added missing manprefix to Makefile.

* Version 1.10 (released 2013-01-31)

  * Changed location of files to /usr/share/yubikey-ksm, etc.

  * Changed location of configuration files to /etc/yubico/ksm/.

  * Fixed bug causing scripts reading the database to fail if
    the config file exists but is not readable by the current user.

* Version 1.9 (released 2013-01-23)

  * Renamed hex2bin to yubi_hex2bin.
    Issue #1 reported by Dain Nilsson, see:
    https://github.com/Yubico/yubikey-ksm/issues/1

* Version 1.8 (released 2013-01-21)

  * Added ChangeLog to releases.

  * Updated documentation for building/installing.

* Version 1.7 (released 2012-12-21)

  * Replaced usage of 'mcrypt_ecb' as it is deprecated.
    Issue #5 reported by Gyula Szabó, see:
    http://code.google.com/p/yubikey-ksm/issues/detail?id=5.

  * Read database config from /etc/yubico/ksm/config-db files,
    as generated by dbconfig.

* Version 1.6 (released 2012-12-18)

  * ykksm-gen-keys supports the PSKC YubiKey profile.

  * ykksm-checksum sorts by serial number first.

  * Removed .pl extension from scripts.

  * Added man pages.

  * Removed ykksm-upgrade.

* Version 1.5 (released 2010-09-14)

  * Brown paper bag release to update NEWS and Makefile before release.

* Version 1.4 (released 2010-09-14)

  * Don't use PDO rowCount, it is not portable.
    Issue #2 reported by arte42.ripe, see:
    http://code.google.com/p/yubikey-ksm/issues/detail?id=2.

  * Fixed perl warning in ykksm-gen-keys.
    Issue #3 reported by toddejohnson, see:
    http://code.google.com/p/yubikey-ksm/issues/detail?id=3.

  * Improve documentation.

* Version 1.3 (released 2010-03-16)

  * Added ykksm-checksum tool.

* Version 1.2 (released 2009-12-15)

  * Documentation and installation experience substantially improved.

* Version 1.1 (released 2009-12-02)

  * Use PHP PDO instead of hard coding use of MySQL database interface.

* Version 1.0 (released 2009-11-19)

  * Initial release.


ykksm-decrypt.php
<?php

# Written by Simon Josefsson <simon@josefsson.org>.
# Copyright (c) 2009-2013 Yubico AB
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are
# met:
#
#   * Redistributions of source code must retain the above copyright
#     notice, this list of conditions and the following disclaimer.
#
#   * Redistributions in binary form must reproduce the above
#     copyright notice, this list of conditions and the following
#     disclaimer in the documentation and/or other materials provided
#     with the distribution.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

ob_start();
require_once 'ykksm-config.php';
require_once 'ykksm-utils.php';
ob_end_clean();

if (isset($_SERVER['HTTP_ACCEPT']) && $_SERVER['HTTP_ACCEPT'] == 'application/json') {
  $format = "json";
 }
else {
  $format = "text";
}

function send_response($status_code, $content, $format) {
  if (!function_exists('http_response_code')) {
    header("X-PHP-Response-Code: $status_code", true, $status_code);
   }
  else {
    http_response_code($status_code);
  }
  if ($format == "json") {
    $content_type = "application/json";
    if (gettype($content) == "string") { $content = array("status" => rtrim($content)); }
    $content = json_encode($content);
   }
  else {
    $content_type = "text/html";
  }
  header("Content-type: $content_type");
  ($status_code == 200) ? print($content) : exit($content);
}


openlog("ykksm", LOG_PID, $logfacility)
  or send_response(500, "ERR Syslog open error\n", $format);

$otp = $_REQUEST["otp"];
if (!$otp) {
  syslog(LOG_INFO, "No OTP provided");
  send_response(400, "ERR No OTP provided\n", $format);
 }

if (!preg_match("/^([cbdefghijklnrtuv]{0,16})([cbdefghijklnrtuv]{32})$/",
		$otp, $matches)) {
  syslog(LOG_INFO, "Invalid OTP format: $otp");
  send_response(400, "ERR Invalid OTP format\n", $format);
 }
$id = $matches[1];
$modhex_ciphertext = $matches[2];

# Oracle support in PDO is highly experimental, OCI is used instead
# Unfortunately PDO and OCI APIs are different...
$use_oci = substr($db_dsn,0,3) === 'oci';

if (!$use_oci) {
  try {
    $dbh = new PDO($db_dsn, $db_username, $db_password, $db_options);
   } catch (PDOException $e) {
    syslog(LOG_ERR, "Database error: " . $e->getMessage());
    send_response(500, "ERR Database error\n", $format);
   }
 }
else {
  # "oci:" prefix needs to be removed before passing db_dsn to OCI
  $db_dsn = substr($db_dsn, 4);
  $dbh = oci_connect($db_username, $db_password, $db_dsn);
  if (!$dbh) {
    $error = oci_error();
    syslog(LOG_err, "Database error: " . $error["message"]);
    send_response(500, "ERR Database error\n", $format);
   }
 }

$sql = "SELECT aeskey, internalname FROM yubikeys " .
       "WHERE publicname = '$id' AND ";

if (!$use_oci) {
  $sql .= "(active OR active = 'true')";
  $result = $dbh->query($sql);
  if (!$result) {
    syslog(LOG_ERR, "Database query error.  Query: " . $sql . " Error: " .
           print_r ($dbh->errorInfo (), true));
    send_response(500, "ERR Database error\n", $format);
   }

  $row = $result->fetch(PDO::FETCH_ASSOC);
  $aeskey = $row['aeskey'];
  $internalname = $row['internalname'];
 }
else {
  $sql .= "active = 1";
  $result = oci_parse($dbh, $sql);
  $execute = oci_execute($result);
  if (!$execute) {
    $error = oci_error($result);
    syslog(LOG_ERR, 'Database query error.   Query:  ' . $sql . 'Error: CODE : ' . $error["code"] .
           ' MESSAGE : ' . $error["message"] . ' POSITION : ' . $error["offset"] .
           ' STATEMENT : ' . $error["sqltext"]);
    send_response(500, "ERR Database error\n", $format);
   }

  $row = oci_fetch_array($result, OCI_ASSOC);
  $aeskey = $row['AESKEY'];
  $internalname = $row['INTERNALNAME'];
 }

if (!$aeskey) {
  syslog(LOG_INFO, "Unknown yubikey: " . $otp);
  send_response(400, "ERR Unknown yubikey\n", $format);
 }

$ciphertext = modhex2hex($modhex_ciphertext);
$plaintext = aes128ecb_decrypt($aeskey, $ciphertext);

$uid = substr($plaintext, 0, 12);
if (strcmp($uid, $internalname) != 0) {
  syslog(LOG_ERR, "UID error: $otp $plaintext: $uid vs $internalname");
  send_response(400, "ERR Corrupt OTP\n", $format);
 }

if (!crc_is_good($plaintext)) {
  syslog(LOG_ERR, "CRC error: $otp: $plaintext");
  send_response(400, "ERR Corrupt OTP\n", $format);
 }

# Mask out interesting fields
$counter = substr($plaintext, 14, 2) . substr($plaintext, 12, 2);
$low = substr($plaintext, 18, 2) . substr($plaintext, 16, 2);
$high = substr($plaintext, 20, 2);
$use = substr($plaintext, 22, 2);

$out = "OK counter=$counter low=$low high=$high use=$use";

syslog(LOG_INFO, "SUCCESS OTP $otp PT $plaintext $out")
  or send_response(500, "ERR Log error\n", $format);

if ($format == "json") {
  send_response(200, array('counter' => $counter, 'low' => $low,
                           'high' => $high, 'use' => $use), $format);
 }
else {
  send_response(200, "$out\n", $format);
 }

# Close database connection.
$dbh = null;

?>
